src/transport/grpc/convert.rs

```diff
@@ -0,0 +1,1349 @@
+        })
+    }
+}
+
+impl TryFrom<Aggregation> for astarte_message_hub_proto::astarte_message::Payload {
+    type Error = MessageHubProtoError;
+
+    fn try_from(value: Aggregation) -> Result<Self, Self::Error> {
+        use astarte_message_hub_proto::astarte_data_type::Data;
+        use astarte_message_hub_proto::astarte_message::Payload;
+
+        let payload = match value {
+            Aggregation::Individual(astarte_type) => {
+                if let AstarteType::Unset = astarte_type {
+                    Payload::AstarteUnset(astarte_message_hub_proto::AstarteUnset {})
+                } else {
+                    let individual_type: astarte_message_hub_proto::AstarteDataTypeIndividual =
+                        astarte_type.try_into()?;
+
+                    Payload::AstarteData(astarte_message_hub_proto::AstarteDataType {
+                        data: Some(Data::AstarteIndividual(individual_type)),
+                    })
+                }
+            }
+            Aggregation::Object(astarte_map) => {
+                let astarte_data: astarte_message_hub_proto::AstarteDataType = astarte_map
+                    .into_iter()
+                    .map(|(k, v)| (k, v.try_into().unwrap()))
+                    .collect::<HashMap<String, astarte_message_hub_proto::AstarteDataTypeIndividual>>()
+                    .into();
+
+                Payload::AstarteData(astarte_data)
+            }
+        };
+
+        Ok(payload)
+    }
+}
+
+impl TryFrom<AstarteType> for astarte_message_hub_proto::astarte_message::Payload {
+    type Error = MessageHubProtoError;
+
+    fn try_from(astarte_device_sdk_type: AstarteType) -> Result<Self, Self::Error> {
+        use astarte_message_hub_proto::astarte_data_type::Data;
+        use astarte_message_hub_proto::astarte_message::Payload;
+        use astarte_message_hub_proto::AstarteDataType;
+
+        let payload = if let AstarteType::Unset = astarte_device_sdk_type {
+            Payload::AstarteUnset(astarte_message_hub_proto::AstarteUnset {})
+        } else {
+            let individual_type: astarte_message_hub_proto::AstarteDataTypeIndividual =
+                astarte_device_sdk_type.try_into()?;
+            Payload::AstarteData(AstarteDataType {
+                data: Some(Data::AstarteIndividual(individual_type)),
+            })
+        };
+        Ok(payload)
+    }
+}
+
+/// This function can be used to convert a map of (String, crate:types::AstarteType) into a
+/// map of (String,  AstarteDataTypeIndividual).
+pub fn map_values_to_astarte_data_type_individual(
+    value: HashMap<String, AstarteType>,
+) -> Result<
+    HashMap<String, astarte_message_hub_proto::AstarteDataTypeIndividual>,
+    MessageHubProtoError,
+> {
+    let mut map: HashMap<String, astarte_message_hub_proto::AstarteDataTypeIndividual> =
+        Default::default();
+    for (key, astarte_type) in value.into_iter() {
+        map.insert(key, astarte_type.try_into()?);
+    }
+    Ok(map)
```

Comment:
Alternatively, you could use the `try_fold` method in this way:


```suggestion
    value.into_iter().try_fold(HashMap::default(), |mut map, (key, astarte_type)| {
            map.insert(key, astarte_type.try_into()?);
            Ok(map)
        })
```

The result should be the same.

Reply:
Refactored the conversion to use `collect` instead to have a pre computed `HashMap` size.